code/edgemedia/src/main/java/com/adobe/marketing/mobile/edge/media/internal/MediaEventProcessor.kt

```diff
@@ -11,18 +11,149 @@
 
 package com.adobe.marketing.mobile.edge.media.internal
 
-internal class MediaEventProcessor {
-    private val SOURCE_TAG = "MediaEventProcessor"
+import androidx.annotation.VisibleForTesting
+import com.adobe.marketing.mobile.Event
+import com.adobe.marketing.mobile.edge.media.internal.MediaInternalConstants.LOG_TAG
+import com.adobe.marketing.mobile.edge.media.internal.xdm.XDMMediaEvent
+import com.adobe.marketing.mobile.services.Log
+import java.util.UUID
 
+internal class MediaEventProcessor(
+    private val mediaState: MediaState,
+    private val dispatcher: (event: Event) -> Unit
+) {
+    private val sourceTag = "MediaEventProcessor"
+    private val sessionsMutex = Any()
+
+    // Determines if a MediaSession can be removed from the queue
+    private val isMediaSessionInactive = { mediaSession: MediaSession ->
+        !mediaSession.isSessionActive && mediaSession.getQueueSize() == 0
+    }
+
+    @VisibleForTesting
+    internal val mediaSessions: MutableMap<String, MediaSession> = mutableMapOf()
+
+    /**
+     * Creates new [MediaSession], assigning it a new random ID.
+     * @return the session ID of the new [MediaSession]
+     */
+    fun createSession(): String {
+        synchronized(sessionsMutex) {
+            val sessionId = UUID.randomUUID().toString()
+            val session = MediaRealTimeSession(sessionId, mediaState, dispatcher)
+            mediaSessions[sessionId] = session
+            Log.trace(LOG_TAG, sourceTag, "Created new session ($sessionId)")
+            return sessionId
+        }
+    }
+
+    /**
+     * Ends the [MediaSession] with the given `sessionId`.
+     * @param sessionId the ID of the [MediaSession] to end
+     * @see [MediaSession.end]
+     */
+    fun endSession(sessionId: String) {
+        synchronized(sessionsMutex) {
+            val session = mediaSessions[sessionId]
+            if (session != null) {
+                session.end()
+                if (isMediaSessionInactive(session)) {
```

Comment:
`endSession()` might not be able to delete session since it might have few events to be sent. That's why I was suggesting we use a clear method which does the inactive check for all the sessions and removes them, like we do for notifyBackendSessionId and notifyErrorResponse methods. That way also don't need extra function var to check since it will be a condition inside that function. And we also have just one way of clearing sessions that is bulk and which will take care of all the sessions which may have been ended but cleared.

Reply:
`endSession()` is passed the sessionId for the session to end, so we don't need to perform a bulk removal like the `notifyX` methods which search for the sessions using the request ID. Here we can just remove this one session using the session id if the session is inactive. The function `isMediaSessionInactive()` checks if the eventQueue is empty so it won't remove a session which still has unprocessed events. 
However, are you suggesting to just call `mediaSessions.values.removeAll { isMediaSessionInactive(it) }` so that any sessions which were previously ended but not removed are also checked?